fdbclient/include/fdbclient/MetaclusterManagement.actor.h

```diff
@@ -1062,6 +1056,329 @@ Future<Void> managementClusterRemoveTenantFromGroup(Transaction tr,
+		if (!tenants.results.empty()) {
+			for (auto t : tenants.results) {
+				self->mgmtClusterTenantMap.emplace(t.first, t.second);
+				lastTenantNameRetrieved = t.first;
+			}
+		}
+
+		return std::pair<bool, TenantName>{ !tenants.more, lastTenantNameRetrieved };
+	}
+
+	ACTOR static Future<std::pair<bool, uint64_t>> getTenantIdNameIndexFromManagementCluster(RestoreClusterImpl* self,
+	                                                                                         Reference<ITransaction> tr,
+	                                                                                         uint64_t initialTenantId) {
+		uint64_t begin = initialTenantId;
+		uint64_t end = std::numeric_limits<uint64_t>::max();
+		state Future<KeyBackedRangeResult<std::pair<int64_t, TenantName>>> tenantsFuture =
+		    ManagementClusterMetadata::tenantMetadata().tenantIdIndex.getRange(
+		        tr, begin, end, CLIENT_KNOBS->MAX_TENANTS_PER_CLUSTER);
+		state KeyBackedRangeResult<std::pair<int64_t, TenantName>> tenants = wait(tenantsFuture);
+
+		state uint64_t lastTenantIdRetrieved;
+		if (!tenants.results.empty()) {
+			for (auto t : tenants.results) {
+				self->mgmtClusterTenantIdIndex.emplace(t.first, t.second);
+				lastTenantIdRetrieved = t.first;
+			}
+		}
+
+		return std::pair<bool, uint64_t>{ !tenants.more, lastTenantIdRetrieved };
+	}
+
+	ACTOR static Future<bool> getTenantsFromManagementClusterForCurrentDataCluster(RestoreClusterImpl* self,
+	                                                                               Reference<ITransaction> tr) {
+		std::pair<Tuple, Tuple> clusterTupleRange =
+		    std::make_pair(Tuple::makeTuple(self->clusterName), Tuple::makeTuple(keyAfter(self->clusterName)));
+		state Future<KeyBackedRangeResult<Tuple>> tenantEntriesFuture =
+		    ManagementClusterMetadata::clusterTenantIndex.getRange(
+		        tr, clusterTupleRange.first, clusterTupleRange.second, CLIENT_KNOBS->MAX_TENANTS_PER_CLUSTER);
+		state KeyBackedRangeResult<Tuple> tenantEntries = wait(tenantEntriesFuture);
+
+		if (!tenantEntries.results.empty()) {
+			for (auto t : tenantEntries.results) {
+				ASSERT(t.getString(0) == self->clusterName);
+				self->mgmtTenantSet.insert(t.getString(1));
+			}
+		}
+
+		return !tenantEntries.more;
+	}
+
+	ACTOR static Future<bool> getAllTenantsFromManagementCluster(RestoreClusterImpl* self) {
+		// get all tenants across all data clusters
+		state TenantName beginRangeTenantName = ""_sr;
+		loop {
+			TenantName initialTenantName = beginRangeTenantName;
+			std::pair<bool, TenantName> tenantsItr = wait(self->ctx.runManagementTransaction(
+			    [self = self, initialTenantName](Reference<typename DB::TransactionT> tr) {
+				    return getTenantsFromManagementCluster(self, tr, initialTenantName);
+			    }));
+
+			if (tenantsItr.first) {
+				// retrieved all the tenants
+				break;
+			} else {
+				// begin range from this tenant
+				beginRangeTenantName = tenantsItr.second;
```

Comment:
Since we already have the returned tenant, you can advance it by one:

```suggestion
				beginRangeTenantName = keyAfter(tenantsItr.second);
```